resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/DriverConfigurationStepsOrchestrator.scala

```diff
@@ -99,40 +99,77 @@ private[spark] class DriverConfigurationStepsOrchestrator(
+            val mountSmallFilesBootstrap = new MountSmallFilesBootstrapImpl(
+                smallFilesSecretName, MOUNTED_SMALL_FILES_SECRET_MOUNT_PATH)
+            val mountSmallLocalFilesStep = new MountSmallLocalFilesStep(
               sparkFiles,
-              jarsDownloadPath,
-              filesDownloadPath,
-              dockerImagePullPolicy,
-              allDriverLabels,
+              smallFilesSecretName,
+              MOUNTED_SMALL_FILES_SECRET_MOUNT_PATH,
+              mountSmallFilesBootstrap)
+            (MOUNTED_SMALL_FILES_SECRET_MOUNT_PATH,
+              sparkFilesWithoutLocal.toArray,
+              Some(mountSmallLocalFilesStep))
+          }
+
+        val initContainerBootstrapStep =
+          if (areAnyFilesNonContainerLocal(sparkJars ++ sparkFilesResolvedFromInitContainer)) {
+            val initContainerConfigurationStepsOrchestrator =
+              new InitContainerConfigurationStepsOrchestrator(
+                namespace,
+                kubernetesResourceNamePrefix,
+                sparkJars,
+                sparkFilesResolvedFromInitContainer,
+                jarsDownloadPath,
+                filesDownloadPath,
+                dockerImagePullPolicy,
+                allDriverLabels,
+                initContainerConfigMapName,
+                INIT_CONTAINER_CONFIG_MAP_KEY,
+                submissionSparkConf)
+            val initContainerConfigurationSteps =
+              initContainerConfigurationStepsOrchestrator.getAllConfigurationSteps()
+            Some(new InitContainerBootstrapStep(initContainerConfigurationSteps,
               initContainerConfigMapName,
-              INIT_CONTAINER_CONFIG_MAP_KEY,
-              submissionSparkConf)
-      val initContainerConfigurationSteps =
-          initContainerConfigurationStepsOrchestrator.getAllConfigurationSteps()
-      Some(new InitContainerBootstrapStep(initContainerConfigurationSteps,
-        initContainerConfigMapName,
-        INIT_CONTAINER_CONFIG_MAP_KEY))
+              INIT_CONTAINER_CONFIG_MAP_KEY))
+        } else Option.empty[DriverConfigurationStep]
+        (submittedLocalFilesDownloadPath,
+          mountSmallFilesWithoutInitContainerStep.toSeq ++
+            initContainerBootstrapStep.toSeq)
     } else {
-      Option.empty[DriverConfigurationStep]
+      (filesDownloadPath, Seq.empty[DriverConfigurationStep])
     }
     val dependencyResolutionStep = new DependencyResolutionStep(
       sparkJars,
       sparkFiles,
       jarsDownloadPath,
-      filesDownloadPath)
+      localFilesDownloadPath)
     Seq(
       initialSubmissionStep,
       kubernetesCredentialsStep,
       dependencyResolutionStep) ++
-      initContainerBootstrapStep.toSeq ++
+      submittedDependenciesBootstrapSteps ++
       pythonStep.toSeq
   }
+
+  private def areAnyFilesNonContainerLocal(files: Seq[String]): Boolean = {
+    files.exists { uri =>
+      Option(Utils.resolveURI(uri).getScheme).getOrElse("file") != "local"
```

Comment:
This might belong in `KubernetesFileUtils` but it has no use other than in here, so I just made it a private method for this.